### PR TITLE
Fleshed out

### PR DESCRIPTION
**Pull request by dastels**

### CODE REVIEW

**Review by kattni — CHANGES_REQUESTED** [review]
Thanks for writing this! 

There's a few issues here, and I'm not sure if they're a cookiecutter failure, a failure on our part to explain cookiecutter's updates clearly, or an issue with what you've entered into cookiecutter. 
* The .files (dot files) are missing, which is hopefully something you missed and not that cookiecutter stopped generating them. Travis isn't running on this because there's no `.travis.yml` file present.
* The module file name is `adafruit_circuitpython_bitbangio` and it should be `adafruit_bitbangio`. Only the repo and the PyPI info should have the full repo-style name in them. 
* Due to the above, there are multiple places where it says `adafruit_circuitpython_bitbangio` and it should say `adafruit_bitbangio` instead (aside from the module file name itself)
* There are still todo's present in multiple files. I won't put details here as they each explain what needs to be done. Do a search for `todo` and you'll find them.

It's late and I haven't looked at much past the cookiecutter issues. This needs to be gone through more thoroughly to make sure that nothing else was missed. Travis will fail in the current state FYI.

Redoing cookiecutter properly and then shifting all the code into where it goes may be the most thorough way to resolve this, however that is a lot of work, and it may also be possible to fix the issues I've found without redoing cookiecutter. I would need more time to go through it and make sure I've located all the things that may be a problem.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      tags: true
+      condition: $DEPLOY_PYPI = "true"
+
+install:
+  - pip install -r requirements.txt
+  - pip install circuitpython-build-tools Sphinx sphinx-rtd-theme
+  - pip install --force-reinstall pylint==1.9.2
+
+script:
+  - pylint adafruit_circuitpython_bitbangio.py
```
Comment (sommersoft) [inline .travis.yml]: change to `- pylint adafruit_bitbangio.py`

File: docs/api.rst [hunk 2]
```diff
@@ -0,0 +1,8 @@
+
+.. If you created a package, create one automodule per module in the package.
+
+.. If your library file(s) are nested in a directory (e.g. /adafruit_foo/foo.py)
+.. use this format as the module name: "adafruit_foo.foo"
+
+.. automodule:: adafruit_circuitpython_bitbangio
```
Comment (sommersoft) [inline docs/api.rst]: change to `.. automodule:: adafruit_bitbangio`

File: docs/conf.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+# Add any paths that contain templates here, relative to this directory.
+templates_path = ['_templates']
+
+source_suffix = '.rst'
+
+# The master toctree document.
+master_doc = 'index'
+
+# General information about the project.
+project = u'Adafruit_circuitpython BitBangIO Library'
```
Comment (sommersoft) [inline docs/conf.py]: change to `project = u'Adafruit BitBangIO Library'`

File: docs/conf.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+     # Latex figure (float) alignment
+     #
+     # 'figure_align': 'htbp',
+}
+
+# Grouping the document tree into LaTeX files. List of tuples
+# (source start file, target name, title,
+#  author, documentclass [howto, manual, or own class]).
+latex_documents = [
+    (master_doc, 'Adafruit_circuitpythonBitBangIOLibrary.tex', u'Adafruit_circuitpythonBitBangIO Library Documentation',
```
Comment (sommersoft) [inline docs/conf.py]: change to `(master_doc, 'AdafruitBitBangIOLibrary.tex', u'Adafruit BitBangIO Library Documentation', `

File: docs/conf.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+# -- Options for Texinfo output -------------------------------------------
+
+# Grouping the document tree into Texinfo files. List of tuples
+# (source start file, target name, title, author,
+#  dir menu entry, description, category)
+texinfo_documents = [
+    (master_doc, 'Adafruit_circuitpythonBitBangIOLibrary', u'Adafruit_circuitpython BitBangIO Library Documentation',
+     author, 'Adafruit_circuitpythonBitBangIOLibrary', 'One line description of project.',
+     'Miscellaneous'),
```
Comment (sommersoft) [inline docs/conf.py:159]: change to:
```
(master_doc, 'AdafruitBitBangIOLibrary', u'Adafruit BitBangIO Library Documentation', 
     author, 'AdafruitBitBangIOLibrary', 'One line description of project.', 
     'Miscellaneous'),

File: docs/conf.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    (master_doc, 'Adafruit_circuitpythonBitBangIOLibrary.tex', u'Adafruit_circuitpythonBitBangIO Library Documentation',
+     author, 'manual'),
+]
+
+# -- Options for manual page output ---------------------------------------
+
+# One entry per manual page. List of tuples
+# (source start file, name, description, authors, manual section).
+man_pages = [
+    (master_doc, 'Adafruit_circuitpythonBitBangIOlibrary', u'Adafruit_circuitpython BitBangIO Library Documentation',
```
Comment (sommersoft) [inline docs/conf.py]: change to `(master_doc, 'AdafruitBitBangIOlibrary', u'Adafruit BitBangIO Library Documentation',`

File: docs/examples.rst [hunk 7]
```diff
@@ -0,0 +1,8 @@
+Simple test
+------------
+
+Ensure your device works with this simple test.
+
+.. literalinclude:: ../examples/bitbangio_simpletest.py
+    :caption: examples/bitbangio_simpletest.py
```
Comment (sommersoft) [inline docs/examples.rst:7]: there is no `examples/bitbangio_simpletest.py` file...

**Review by sommersoft — CHANGES_REQUESTED** [review]
I went ahead and denoted the additional areas that need the name changes @kattni mentioned. Like she said, it may save time to just run the cookiecutter again and move over the library files. And please, please, please let us know if we can tweak the "new" process to make it better/easier.

With regard to the `bme280_bbspi` examples, I understand that the current library needed to be changed since `adafruit_bus_device` is hardcoded in. However, I don't think it serves as a good example, nor should it require a user (or anyone else) to re-write and maintain a "bitbang" version of the library. I don't have a full solution in my mind how this can be addressed across the library landscape. Probably a good thing to bring up `In The Weeds` during a meeting.

I'm also thinking that `SPI` and `SPIDevice` can be sub/super classed, but that may just be a preference.

Looks good though! And, I imagine it was a fun one to work out.. :smile:

**Comment by sommersoft** [comment]
argh! A few of my comments dropped. Let's see if I can remember them. 🤦‍♂️

File: adafruit_bitbangio.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_BitBangIO.git"
+
+import time
+import digitalio
+
+class SPI(object):
+
+    def __init__(self, clock=None, MOSI=None, MISO=None):
+        if clock is None or MOSI is None or MISO is None:
+            return None
```
Comment (sommersoft) [inline adafruit_bitbangio.py]: Would it be better to raise an exception here? Returning `None` still creates the `SPI` object, but doesn't initialize anything. Further, `SPIDevice` doesn't check for a "valid" `SPI` object (i.e. `not None`) and continues on to initialize `cs` and `spi.configure`.

I recommend raising `ValueError`, denoting all 3 pins are required.

Example of what I'm saying:
```
Python 3.6.2 (v3.6.2:5fd33b5, Jul  8 2017, 04:57:36) [MSC v.1900 64 bit (AMD64)] on win32
Type "copyright", "credits" or "license()" for more information.
>>> class SPI(object):
	def __init__(self, clock=None, MOSI=None):
		if clock is None or MOSI is None:
			return None
		self._clock = clock
		self._MOSI = MOSI

	def values(self):
		return self._clock, self._MOSI

>>> class SPIDevice(object):
	def __init__(self, spi):
		self._spi = spi

	def spi(self):
		return self._spi

>>> x = SPI()
>>> y = SPIDevice(x)
>>> y.spi
<__main__.SPI object at 0x000001EA656A6E48>
>>> y.spi.values
Traceback (most recent call last):
  File "<pyshell#13>", line 1, in <module>
    y.spi.values
  File "<pyshell#2>", line 9, in values
    return self._clock, self._MOSI
AttributeError: 'SPI' object has no attribute '_clock'
>>> 
```

File: adafruit_bitbangio.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        """Attempts to grab the SPI lock. Returns True on success.
+
+        :return: True when lock has been grabbed
+        """
+        return True
+
+
+    def unlock(self):
+        """Releases the SPI lock."""
+        pass
```
Comment (sommersoft) [inline adafruit_bitbangio.py:114]: Is there a way to utilize Blinka's [Lockable class](https://github.com/adafruit/Adafruit_Blinka/blob/master/src/adafruit_blinka/__init__.py#L53) for `try_lock` and `unlock`? I have zero experience with Blinka yet, so this is more of a question than a suggestion.

**Review by sommersoft — CHANGES_REQUESTED** [review]
Missing Comments Found!

File: examples/bme280_bbspi.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
+import time
+from adafruit_bitbangio import SPIDevice
```
Comment (ladyada) [inline examples/bme280_bbspi.py:2]: we shouldn't need this right? if bitbangio.SPI replaces only busio.SPI then you should be able to swap in either

**Review by ladyada — CHANGES_REQUESTED** [review]

File: examples/bme280_bbspi.py (hunk 10 above)
Comment (dastels) [reply examples/bme280_bbspi.py:2]: "import adafruit_bus_device.spi_device as spi_device"

is hardcoded into Adafruit_CircuitPython_BME280/adafruit_bme280.py : Adafruit_BME280_SPI

Comment (dastels) [reply examples/bme280_bbspi.py:2]: I suppose we could subclass Adafruit_CircuitPython_BME280/adafruit_bme280.py : Adafruit_BME280 with something that works with the bitbanged support.

File: adafruit_bitbangio.py (hunk 8 above)
Comment (dastels) [reply adafruit_bitbangio.py]: Done

File: examples/bme280_bbspi.py (hunk 10 above)
Comment (ladyada) [reply examples/bme280_bbspi.py:2]: why would you have to? bitbangio.SPI should be completely identical API to the hardware SPI - whats different?

Comment (ladyada) [reply examples/bme280_bbspi.py:2]: (we pass SPI into spi_device - the spi_device part doesnt do any hardware SPI stuff)

**Comment by dastels** [comment]
Here's my thinking after poking around a bit.
Using adafruit_bus_device:SPIDevice looks like it should work. I'll give it a try next.
Adafruit_BME280:BME280 only supports I2C.
I was using a BME280 class based on the circuitpython version, which subclasses for I2C and SPI. That wasn't working and I eventually tracked it down to it's _read_register method which does a write of the register number, followed by a readinto of the data. Changing that to use write_readinto to do everythign in one transaction worked.

File: .travis.yml (hunk 1 above)
Comment (dastels) [reply .travis.yml]: Agreed. It isn't a great example (not very simple), but it's what I had... I used it as a case to test the implementation.

Also agreed about rewriting. The existing drivers weren't written with the idea of swappable SPI implementations in mind. Maybe a slight rearchitecting is in order to make it all more modular?

**Comment by kattni** [comment]
@dastels Are you still working on this?

**Comment by kattni** [comment]
I'm closing this PR for now. @dastels If you are interested in returning to it, please let me know and I can reopen it.

**Pull request closed without merging**